system/jlib/jmutex.hpp

```diff
@@ -1042,4 +1042,117 @@ class Singleton
     CriticalSection cs;
 };
 
+// Similar to class Shared<X>, but thread safe versions of the functions (avoid the need for critical sections)
+// Optional atomic query(std::function<CLASS *()> factoryFunc, CriticalSection & cs) allows for a singleton initialization.
+template <class CLASS> class AtomicShared
+{
+public:
+    inline AtomicShared()                              { ptr.store(nullptr, std::memory_order_relaxed); }
+    inline AtomicShared(CLASS * _ptr, bool owned)      { ptr.store(_ptr, std::memory_order_relaxed); if (!owned && _ptr) _ptr->Link(); }
+    inline AtomicShared(const AtomicShared & other)    { ptr.store(other.getLinkNonAtomic(), std::memory_order_relaxed); }
+#if defined(__cplusplus) && __cplusplus >= 201100
+    inline AtomicShared(AtomicShared && other)         { ptr.store(other.getClear(), std::memory_order_relaxed); }
+#endif
+    inline ~AtomicShared()                             { ::Release(ptr.load(std::memory_order_relaxed)); }
+    inline AtomicShared<CLASS> & operator = (const AtomicShared<CLASS> & other) { this->setown(other.getLinkNonAtomic()); return *this;  }
+
+    inline void clear()                                { ::Release(getClear()); }
+    inline CLASS * getClear()
+    {
+        return ptr.exchange(nullptr);
+    }
+    inline CLASS * getClearNonAtomic()
+    {
+        CLASS * result = ptr.load();
+        ptr.store(nullptr);
+        return result;
+    }
+
+    //The getLink() function cannot be implemented in a thread safe way - e.g. if clear is called concurrently
+    //then temp will point to a freed object.  (Might be possible with support for transactional memory...)
+    inline CLASS * getLinkNonAtomic() const
+    {
+        CLASS * temp = ptr;
+        if (temp)
+            temp->Link();
+        return temp;
+    }
+    inline CLASS * query() const
+    {
+        return ptr.load(std::memory_order_acquire);
+    }
+    inline CLASS * query(std::function<CLASS *()> factoryFunc, CriticalSection & cs)
```

Comment:
I wondered, so I checked...  The overhead for passing the std::function is actually quite high (i.e. even if filled in) .  **Much** better code generated with

```
    template <typename FUNC>
    inline CLASS * query(FUNC factoryFunc, CriticalSection & cs)

```

Reply:
hm ok, I'll revert to template method